Panic during VTGATE shutdown

### Overview of the Issue

During a VTGate compute rebalance, we frequently observe the following event in Up:

> “Service crashed while stopping”

On inspecting the container logs, we consistently find the following panic:

`
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x189420c]

goroutine 1063399 [running]:
vitess.io/vitess/go/vt/topo/etcd2topo.(*Server).Get(0xc00092a1c0, {0x232a0f8, 0xc006500bd0}, {0xc001e2f2f0?, 0x29?})
    vitess.io/vitess/go/vt/topo/etcd2topo/file.go:80 +0x8c
    vitess.io/vitess/go/vt/topo.(*StatsConn).Get(0xc000920780, {0x232a0f8, 0xc006500bd0}, {0xc001e2f2f0, 0x2e})
        vitess.io/vitess/go/vt/topo/stats_conn.go:110 +0x18b
        vitess.io/vitess/go/vt/topo.(*Server).GetShard(0xc00092c240, {0x232a130, 0xc005da3540}, {0xc00005dae8, 0x11}, {0xc0016c72b8, 0x5})
            vitess.io/vitess/go/vt/topo/shard.go:184 +0x2b0
            vitess.io/vitess/go/vt/discovery.(*keyspaceState).getMoveTablesStatus.func1()
                vitess.io/vitess/go/vt/discovery/keyspace_events.go:485 +0x6d
                golang.org/x/sync/errgroup.(*Group).Go.func1()
                    golang.org/x/sync@v0.10.0/errgroup/errgroup.go:78 +0x50
                    created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 4361
                        golang.org/x/sync@v0.10.0/errgroup/errgroup.go:75 +0x96
`
This occurs in the context of a service shutdown / host rebalance scenario.



### Reproduction Steps

This error is not 100% deterministic it happens due to a race condition between:

1. Shutdown of the topo server (etcd client set to nil)
2. Goroutines still using the topo server (e.g. event watchers)

Steps:
1. Run vtgate on some N instances.
2. Shutdown vtgate on a few instances.
3. Check logs of the vtgate host and you'll find these errors.

### Binary Version

```sh
vtgate version Version: 21.0.3-SNAPSHOT (Git revision 933f080eb3c6e8c551cf9591029c1e2fb94317b0 branch 'sharded/revert-schema-shard-by-shard') built on Fri Jun  6 13:35:54 IST 2025 by siddharth.singh@siddharth-singh-DM6YX2DV71 using go1.24.1 darwin/arm64
```

### Operating System and Environment details

```sh
Kernel: Darwin 24.5.0
Architecture: ARM
```

### Log Fragments

```sh

```

--- a/go/vt/topo/etcd2topo/file.go
+++ b/go/vt/topo/etcd2topo/file.go
@@ -28,6 +28,9 @@ import (
 
 // Create is part of the topo.Conn interface.
 func (s *Server) Create(ctx context.Context, filePath string, contents []byte) (topo.Version, error) {
+	if err := s.checkClosed(); err != nil {
+		return nil, convertError(err, filePath)
+	}
 	nodePath := path.Join(s.root, filePath)
 
 	// We have to do a transaction, comparing existing version with 0.
@@ -47,6 +50,9 @@ func (s *Server) Create(ctx context.Context, filePath string, contents []byte) (
 
 // Update is part of the topo.Conn interface.
 func (s *Server) Update(ctx context.Context, filePath string, contents []byte, version topo.Version) (topo.Version, error) {
+	if err := s.checkClosed(); err != nil {
+		return nil, convertError(err, filePath)
+	}
 	nodePath := path.Join(s.root, filePath)
 
 	if version != nil {
@@ -75,6 +81,9 @@ func (s *Server) Update(ctx context.Context, filePath string, contents []byte, v
 
 // Get is part of the topo.Conn interface.
 func (s *Server) Get(ctx context.Context, filePath string) ([]byte, topo.Version, error) {
+	if err := s.checkClosed(); err != nil {
+		return nil, nil, convertError(err, filePath)
+	}
 	nodePath := path.Join(s.root, filePath)
 
 	resp, err := s.cli.Get(ctx, nodePath)
@@ -90,6 +99,9 @@ func (s *Server) Get(ctx context.Context, filePath string) ([]byte, topo.Version
 
 // GetVersion is part of the topo.Conn interface.
 func (s *Server) GetVersion(ctx context.Context, filePath string, version int64) ([]byte, error) {
+	if err := s.checkClosed(); err != nil {
+		return nil, convertError(err, filePath)
+	}
 	nodePath := path.Join(s.root, filePath)
 
 	resp, err := s.cli.Get(ctx, nodePath, clientv3.WithRev(version))
@@ -105,6 +117,9 @@ func (s *Server) GetVersion(ctx context.Context, filePath string, version int64)
 
 // List is part of the topo.Conn interface.
 func (s *Server) List(ctx context.Context, filePathPrefix string) ([]topo.KVInfo, error) {
+	if err := s.checkClosed(); err != nil {
+		return []topo.KVInfo{}, convertError(err, filePathPrefix)
+	}
 	nodePathPrefix := path.Join(s.root, filePathPrefix)
 
 	resp, err := s.cli.Get(ctx, nodePathPrefix, clientv3.WithPrefix())
@@ -127,6 +142,9 @@ func (s *Server) List(ctx context.Context, filePathPrefix string) ([]topo.KVInfo
 
 // Delete is part of the topo.Conn interface.
 func (s *Server) Delete(ctx context.Context, filePath string, version topo.Version) error {
+	if err := s.checkClosed(); err != nil {
+		return convertError(err, filePath)
+	}
 	nodePath := path.Join(s.root, filePath)
 
 	if version != nil {

--- a/go/vt/topo/etcd2topo/server.go
+++ b/go/vt/topo/etcd2topo/server.go
@@ -34,6 +34,7 @@ We follow these conventions within this package:
 package etcd2topo
 
 import (
+	"context"
 	"crypto/tls"
 	"crypto/x509"
 	"strings"
@@ -91,6 +92,16 @@ func registerEtcd2TopoFlags(fs *pflag.FlagSet) {
 	utils.SetFlagStringVar(fs, &clientCertPath, "topo-etcd-tls-cert", clientCertPath, "path to the client cert to use to connect to the etcd topo server, requires topo-etcd-tls-key, enables TLS")
 	utils.SetFlagStringVar(fs, &clientKeyPath, "topo-etcd-tls-key", clientKeyPath, "path to the client key to use to connect to the etcd topo server, enables TLS")
 	utils.SetFlagStringVar(fs, &serverCaPath, "topo-etcd-tls-ca", serverCaPath, "path to the ca to use to validate the server cert when connecting to the etcd topo server")
+}
+
+// checkClosed returns context.Canceled if the server has been closed.
+// This mimics the pattern used for context cancellation which gets converted
+// to topo.Interrupted by convertError().
+func (s *Server) checkClosed() error {
+	if s.cli == nil {
+		return context.Canceled
+	}
+	return nil
 }
 
 // Close implements topo.Server.Close.

--- a/go/vt/topo/etcd2topo/server_test.go
+++ b/go/vt/topo/etcd2topo/server_test.go
@@ -361,6 +361,75 @@ func TestEtcd2TopoGetTabletsPartialResults(t *testing.T) {
 	}
 }
 
+// TestEtcd2TopoServerClosed tests that operations on a closed server return
+// appropriate errors instead of panicking due to nil pointer dereference.
+func TestEtcd2TopoServerClosed(t *testing.T) {
+	// Start a single etcd in the background.
+	clientAddr, _ := startEtcd(t, 0)
+
+	testRoot := "/test-closed"
+
+	// Create the server on the new root.
+	ts, err := topo.OpenServer("etcd2", clientAddr, path.Join(testRoot, topo.GlobalCell))
+	require.NoError(t, err, "OpenServer() failed: %v", err)
+
+	// Create the CellInfo first.
+	ctx := context.Background()
+	err = ts.CreateCellInfo(ctx, "test_cell", &topodatapb.CellInfo{
+		ServerAddress: clientAddr,
+		Root:          path.Join(testRoot, "test_cell"),
+	})
+	require.NoError(t, err, "CreateCellInfo() failed: %v", err)
+
+	// Get the connection for the cell
+	conn, err := ts.ConnForCell(ctx, "test_cell")
+	require.NoError(t, err, "ConnForCell() failed: %v", err)
+
+	// Test that operations work before closing
+	testPath := "test_key"
+	testContents := []byte("test_value")
+
+	_, err = conn.Create(ctx, testPath, testContents)
+	require.NoError(t, err, "Create() before close should succeed")
+
+	// Close the connection
+	ts.Close()
+
+	// Test that operations return appropriate errors after closing
+	_, err = conn.Create(ctx, "another_key", testContents)
+	require.Error(t, err, "Create() after close should fail")
+	require.True(t, topo.IsErrType(err, topo.Interrupted), "Error should be topo.Interrupted, got: %v", err)
+
+	_, _, err = conn.Get(ctx, testPath)
+	require.Error(t, err, "Get() after close should fail")
+	require.True(t, topo.IsErrType(err, topo.Interrupted), "Error should be topo.Interrupted, got: %v", err)
+
+	_, err = conn.GetVersion(ctx, testPath, 1)
+	require.Error(t, err, "GetVersion() after close should fail")
+	require.True(t, topo.IsErrType(err, topo.Interrupted), "Error should be topo.Interrupted, got: %v", err)
+
+	err = conn.Delete(ctx, testPath, nil)
+	require.Error(t, err, "Delete() after close should fail")
+	require.True(t, topo.IsErrType(err, topo.Interrupted), "Error should be topo.Interrupted, got: %v", err)
+
+	_, err = conn.List(ctx, "/")
+	require.Error(t, err, "List() after close should fail")
+	require.True(t, topo.IsErrType(err, topo.Interrupted), "Error should be topo.Interrupted, got: %v", err)
+
+	_, err = conn.Update(ctx, testPath, testContents, nil)
+	require.Error(t, err, "Update() after close should fail")
+	require.True(t, topo.IsErrType(err, topo.Interrupted), "Error should be topo.Interrupted, got: %v", err)
+
+	// Test watch operations after close
+	_, _, err = conn.Watch(ctx, testPath)
+	require.Error(t, err, "Watch() after close should fail")
+	require.True(t, topo.IsErrType(err, topo.Interrupted), "Error should be topo.Interrupted, got: %v", err)
+
+	_, _, err = conn.WatchRecursive(ctx, "/")
+	require.Error(t, err, "WatchRecursive() after close should fail")
+	require.True(t, topo.IsErrType(err, topo.Interrupted), "Error should be topo.Interrupted, got: %v", err)
+}
+
 // testKeyspaceLock tests etcd-specific heartbeat (TTL).
 // Note TTL granularity is in seconds, even though the API uses time.Duration.
 // So we have to wait a long time in these tests.

--- a/go/vt/topo/etcd2topo/watch.go
+++ b/go/vt/topo/etcd2topo/watch.go
@@ -34,6 +34,9 @@ import (
 
 // Watch is part of the topo.Conn interface.
 func (s *Server) Watch(ctx context.Context, filePath string) (*topo.WatchData, <-chan *topo.WatchData, error) {
+	if err := s.checkClosed(); err != nil {
+		return nil, nil, convertError(err, filePath)
+	}
 	nodePath := path.Join(s.root, filePath)
 
 	// Get the initial version of the file
@@ -160,6 +163,9 @@ func (s *Server) Watch(ctx context.Context, filePath string) (*topo.WatchData, <
 
 // WatchRecursive is part of the topo.Conn interface.
 func (s *Server) WatchRecursive(ctx context.Context, dirpath string) ([]*topo.WatchDataRecursive, <-chan *topo.WatchDataRecursive, error) {
+	if err := s.checkClosed(); err != nil {
+		return nil, nil, convertError(err, dirpath)
+	}
 	nodePath := path.Join(s.root, dirpath)
 	if !strings.HasSuffix(nodePath, "/") {
 		nodePath = nodePath + "/"

